### ISSUE TITLE
Very poor performance of Pagination.{pages,iter_pages}

### ISSUE THREAD

**Issue by wodim** [pallets-eco/flask-sqlalchemy#622]
The performance of those two methods is really really bad when you have thousands of pages. In my case, with 100k pages, painting the paginator takes over 300 ms.

`Pagination.iter_pages` is slow by itself. `Pagination.pages` is not that slow but it's called once for every page from `Pagination.iter_pages` when you consume the iterator, and calling it 100k times really adds up. Python is a very slow language and calling the same method so many times is undesirable even if [the method seems simple](https://github.com/mitsuhiko/flask-sqlalchemy/blob/381653489af8ee8aa26340a7c3fe0a0dced64258/flask_sqlalchemy/__init__.py#L325-L332).

The low-hanging fruit here is caching the result of `Pagination.pages`. That fixes more than half of the problem (only the call to `Pagination.iter_pages` remains). I did so with this subclass, which might be helpful to someone. (I'm far from an expert in Python, so this may hurt your eyes)

```python
class MyPagination(Pagination):

    def __init__(self, query, page, per_page, total, items):
        super().__init__(query, page, per_page, total, items)
        self._pages = super().pages

    @property
    def pages(self):
        return self._pages
```

But of course a better solution is desirable.

**Comment by dourvaris** [comment]
https://github.com/mitsuhiko/flask-sqlalchemy/pull/393

**Comment by davidism** [comment]
fixed in #1087
